### PR TITLE
Avoid usage of MemoryMarshal.Cast in Marvin

### PR DESCRIPTION
**Pull request by kasper3**
Based on the discussion from https://github.com/dotnet/corefx/pull/25735#discussion_r155290532

### CODE REVIEW

**Comment by jkotas** [comment]
Some measurements on why does this help?

Note that there are multiple places where the non-portable Cast is used.

**Comment by kasper3** [comment]
@jkotas, I will try to measure the diff a bit later on a bit faster system than I am currently on. 😄

File: src/Common/src/System/Marvin.cs [hunk 1]
```diff
@@ -57,7 +57,7 @@ public static long ComputeHash(ReadOnlySpan<byte> data, ulong seed)
                     break;
 
                 case 2:
-                    p0 += 0x800000u | MemoryMarshal.Cast<byte, ushort>(data)[0];
+                    p0 += 0x800000u | ((uint)data[0]) << 8 | data[1];
```
Comment (krwq) [inline src/Common/src/System/Marvin.cs]: there is also one more: https://github.com/dotnet/corefx/pull/28110/files#diff-cd635355a880d8cef033635b10e2e4deR64

Comment (krwq) [inline src/Common/src/System/Marvin.cs]: also please use brackets as operator precedence is not obvious

Comment (kasper3) [reply src/Common/src/System/Marvin.cs]: Updated in the followup push.

**Comment by mikedn** [comment]
I'm not sure I understand the point of this change, `MemoryMarshal.Cast` was "fixed" in https://github.com/dotnet/coreclr/pull/16659. This change would be useful if alignment is a concern but as far as performance is concerned having 2 memory accesses instead of 1 is unlikely to be a win.

**Comment by krwq** [comment]
I thought the original plan was to not fix MemoryMarshal.Cast (PR recommendation was to change it as in this PR) but since it is fixed I agree we should not merge it unless there is some perf data behind it (which we probably should investigate rather than merge)

**Comment by kasper3** [comment]
@krwq, https://github.com/dotnet/corefx/issues/27786 has some steps to reproduce the current-regression. On Linux, 2.1 has regression compared to 2.0.

**Pull request closed without merging**